File: implementations/csharp/Provekit.IR/Term.cs
// SPDX-License-Identifier: Apache-2.0
//
// IR-JSON v1.1.0 Term. Three kinds: VarTerm (no sort, named only),
// ConstTerm (carries sort + value), CtorTerm (no sort, named with args).
// Mirrors implementations/rust/provekit-ir-symbolic/src/lib.rs and
// implementations/cpp/provekit-ir-symbolic/include/provekit/ir.hpp.

namespace Provekit.IR;

public abstract record Term;

public sealed record VarTerm(string Name) : Term;

/// <summary>Const value variant — Int / String / Bool.</summary>
public abstract record ConstValue
{
    public sealed record Int(long Value) : ConstValue;
    public sealed record Str(string Value) : ConstValue;
    public sealed record Bool(bool Value) : ConstValue;
}

public sealed record ConstTerm(ConstValue Value, Sort Sort) : Term;

public sealed record CtorTerm(string Name, IReadOnlyList<Term> Args) : Term;

public sealed record LambdaTerm(string ParamName, Sort ParamSort, Term Body) : Term;

public sealed record LetBinding(string Name, Term BoundTerm);

public sealed record LetTerm(IReadOnlyList<LetBinding> Bindings, Term Body) : Term;

public static class Terms
{
    public static Term Var(string name) => new VarTerm(name);

    public static Term Num(long value) =>
        new ConstTerm(new ConstValue.Int(value), Sort.Int);

    public static Term StrConst(string value) =>
        new ConstTerm(new ConstValue.Str(value), Sort.String);

    public static Term BoolConst(bool value) =>
        new ConstTerm(new ConstValue.Bool(value), Sort.Bool);

    public static Term Ctor(string name, params Term[] args) =>
        new CtorTerm(name, args);

    public static Term Lambda(string paramName, Sort paramSort, Term body) =>
        new LambdaTerm(paramName, paramSort, body);

    public static LetBinding Binding(string name, Term boundTerm) =>
        new LetBinding(name, boundTerm);

    public static Term Let(IReadOnlyList<LetBinding> bindings, Term body) =>
        new LetTerm(bindings, body);

    /// <summary>
    /// References the return value within a post formula. Compiles to a
    /// VarTerm whose name matches the enclosing contract's outBinding
    /// (default "out"). Custom outBindings can use <see cref="Var"/>
    /// directly.
    /// </summary>
    public static Term Out() => new VarTerm("out");
}

File: implementations/java/provekit-ir/src/main/java/com/provekit/ir/Term.java
package com.provekit.ir;

/**
 * Sealed term hierarchy matching the CDDL spec.
 */
public sealed interface Term {
    String toJson();

    record Var(String name, Sort sort) implements Term {
        public String toJson() {
            return "{\"kind\":\"var\",\"name\":\"" + Sort.escape(name) + "\"}";
        }
    }

    record Const(Value value, Sort sort) implements Term {
        public String toJson() {
            return "{\"kind\":\"const\",\"sort\":" + sort.toJson() + ",\"value\":" + value.toJson() + "}";
        }
    }

    record Ctor(String name, Term[] args, Sort sort) implements Term {
        public String toJson() {
            StringBuilder sb = new StringBuilder("{\"args\":[");
            for (int i = 0; i < args.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(args[i].toJson());
            }
            sb.append("],\"kind\":\"ctor\",\"name\":\"").append(Sort.escape(name)).append("\"}");
            return sb.toString();
        }
    }

    record Lambda(String paramName, Sort paramSort, Term body, Sort sort) implements Term {
        public String toJson() {
            return "{\"kind\":\"lambda\",\"paramName\":\"" + Sort.escape(paramName) + "\",\"paramSort\":" + paramSort.toJson() + ",\"body\":" + body.toJson() + "}";
        }
    }

    record Let(LetBinding[] bindings, Term body, Sort sort) implements Term {
        public String toJson() {
            StringBuilder sb = new StringBuilder("{\"kind\":\"let\",\"bindings\":[");
            for (int i = 0; i < bindings.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(bindings[i].toJson());
            }
            sb.append("],\"body\":").append(body.toJson()).append("}");
            return sb.toString();
        }
    }

    record LetBinding(String name, Term boundTerm) {
        public String toJson() {
            return "{\"name\":\"" + Sort.escape(name) + "\",\"boundTerm\":" + boundTerm.toJson() + "}";
        }
    }

    sealed interface Value {
        String toJson();
        record Int(long value) implements Value {
            public String toJson() { return String.valueOf(value); }
        }
        record Str(String value) implements Value {
            public String toJson() { return "\"" + Sort.escape(value) + "\""; }
        }
        record Bool(boolean value) implements Value {
            public String toJson() { return String.valueOf(value); }
        }
        record Real(double value) implements Value {
            public String toJson() { return String.valueOf(value); }
        }
    }

    // Convenience constructors
    static Term var_(String name, Sort sort) { return new Var(name, sort); }
    static Term var_(String name) { return new Var(name, Sort.Ref); }
    static Term const_(long value, Sort sort) { return new Const(new Value.Int(value), sort); }
    static Term const_(String value, Sort sort) { return new Const(new Value.Str(value), sort); }
    static Term const_(boolean value, Sort sort) { return new Const(new Value.Bool(value), sort); }
    static Term const_(double value, Sort sort) { return new Const(new Value.Real(value), sort); }
    static Term ctor(String name, Term[] args, Sort sort) { return new Ctor(name, args, sort); }
    static Term lambda(String paramName, Sort paramSort, Term body, Sort sort) { return new Lambda(paramName, paramSort, body, sort); }
    static Term let(LetBinding[] bindings, Term body, Sort sort) { return new Let(bindings, body, sort); }
    static LetBinding binding(String name, Term boundTerm) { return new LetBinding(name, boundTerm); }
}
